Add documentation of binprecice (#115)
precice/precice.github.io: pages/docs/tooling/tooling-builtin.md

---
title: Built-in tooling
permalink: tooling-builtin.html
keywords: tooling, xml, configuration, version
summary: "Built-in tooling is always installed alongside preCICE and provides some basic functionality."
---

Part of a preCICE installation is the tool `binprecice`.
It provides an easy-to-use interface to tooling API of the preCICE library.

## XML reference

```bash
binprecice md
binprecice xml
binprecice dtd
```

> This prints the XML reference to the console in various flavours.

Reviewer: ```suggestion
This prints the XML reference to the console in various flavors.
```
:grin:

Author: Oh boy

Be aware that this version does not contain styling, LaTeX rendering, and functioning links.

```bash
binprecice md | markdown > reference.html
```

### `xml`

This prints the XML configuration with in-lined annotations of tags and attributes.

### `dtd`

This prints the DTD information, which can be used to validate the XML configuration file.

## preCICE version

{% version 2.4.0 %}
This feature is available since version 2.4.0.
{% endversion %}

```bash
binprecice version
```

This prints the version information of preCICE, which consists of multiple semicolon-separated parts.

1. the version of preCICE e.g. `2.3.0`.
2. the git revision of this version e.g. `v2.3.0-87-g04ee7308-dirty`.
   This is interesting for development versions of preCICE.
   It is not available if the library wasn't build using the git repository.
3. Configuration options for MPI, PETSc, Python and some more.
4. Compilation and linker flags used to build preCICE

## Configuration check

{% version 2.4.0 %}
This feature is available since version 2.4.0.
{% endversion %}

```bash
binprecice check FILE [ PARTICIPANT [ COMMSIZE ] ]
```

> The `check` runs the preCICE configuration parsing and checking logic on the given configuration file.

Reviewer: We have to be careful to not oversell the feature here. A user might think that the call also checks whether the exchange tags are set correctly or similar. It "only" validates whether the configuration follows the semantics, not the "logic" of  the configuration.

Reviewer 2: Maybe also give an example of issues it can detect and an example of issues it cannot detect?

Author: True, some more information would be appropriate here.

> A user might think that the call also checks whether the exchange tags are set correctly or similar. It "only" validates whether the configuration follows the semantics, not the "logic" of the configuration.

This is a bit wishy washy at the moment and mostly due to how our configuration works.

As an example the config parser checks if the participants of the exchange tags are part of the cplscheme and whether they hold the data, but doesn't ensure that all required data is transferred.
It checks "most of the local logic", if that makes sense.

Author: I added some clarifications

This will find the majority of the configuration mistakes without having to start a simulation.

The basic usage is to simply check a configuration file.

```bash
binprecice check precice.xml
```

To enable more checks, additionally pass the name of one participant.
This should be repeated for every defined participant in the simulation.

```bash
binprecice check precice.xml Fluid
```

> To enable all available checks, you may additionally pass the communicator size of the Participant.

Reviewer: ```suggestion
To enable all available checks, you may additionally pass the communicator size of the participant.
```

> 

Reviewer 2: As part of this PR, maybe also extend the introduction a bit. "an easy-to-use interface to tooling API" is a bit vague for the user.

Maybe write something like:

```suggestion

With `binprecice`, you can get the installed preCICE version, generate a reference of all available configuration options, as well as check your configuration file for basic configuration issues.
```

Author: Good point

3. Configuration options for MPI, PETSc, Python and some more.
4. Compilation and linker flags used to build preCICE

## Configuration check

{% version 2.4.0 %}
This feature is available since version 2.4.0.
{% endversion %}

```bash
binprecice check FILE [ PARTICIPANT [ COMMSIZE ] ]
```

The `check` runs the preCICE configuration parsing and checking logic on the given configuration file.
This will find the majority of the configuration mistakes without having to start a simulation.

The basic usage is to simply check a configuration file.

```bash
binprecice check precice.xml
```

To enable more checks, additionally pass the name of one participant.
This should be repeated for every defined participant in the simulation.

```bash
binprecice check precice.xml Fluid
```

To enable all available checks, you may additionally pass the communicator size of the Participant.

> This enables some niche checks, which should not be necessary in the vast majority of cases.

Reviewer 2: While I do like that the "niche checks" tells me this is probably not something I need, I think that this is also a bit empty of additional information. Some examples / clarification in parentheses would help, otherwise every time I have a problem I would just enable everything without thinking.

Author: I rewrote the section for clarity.

The `check` runs the preCICE configuration parsing and checking logic on the given configuration file.
This will find the majority of the configuration mistakes without having to start a simulation.
These checks include wrong tags and attribute values, and more elaborate naming checks.

> More intricate logic, such as checks if all necessary data are exchanged in a Coupling Schemes, are not covered.

Reviewer 2: ```suggestion
More advanced logic, such as checks if all necessary data are exchanged in a coupling scheme, are not covered.
```

(we should use rather simpler vocabulary in documentation, as the goal is clarity and the target audience mostly non-native speakers)

The basic usage is to simply check a configuration file:

```bash

> binprecice check precice.xml

Reviewer 2: ```suggestion
binprecice check precice-config.xml
```
(common name)

  ```log
  ERROR: The configuration contains an unknown tag <data:vektor>.
  ```

* Misspelled data names (should be `Forces` instead)

  ```log
  ERROR: Data with name "forces" used by mesh "Solid" is not defined. Please define a data tag with name="forces".
  ```

* Incorrect attribute combinations (mesh provided and received at the same time)

  ```log
  ERROR: Participant "SolverOne" cannot receive and provide mesh "Test-Square" at the same time. Please remove all but one of the "from" and "provide" attributes in the <use-mesh name="Test-Square"/> node of SolverOne.
  ```

* Incorrect meshes used in mapping definitions (`MeshTwo` doesn't exist)

  ```log
  ERROR: Mesh "MeshTwo" was not found while creating a mapping. Please correct the to="MeshTwo" attribute.
  ```

To enable more niche checks, additionally pass the name of one participant.
This participant is assumed to run on a single rank.
You may additionally pass the communicator size of the participant.
This enables some checks regarding user-defined intra-participant communication, which should not be necessary in the vast majority of cases.

```bash
binprecice check precice.xml Fluid

> binprecice check precice.xml Fluid 2

Reviewer 2: ```suggestion
binprecice check precice-config.xml Fluid
binprecice check precice-config.xml Fluid 2
```
